lib/puppet/ssl/host.rb

```diff
@@ -419,6 +431,166 @@ def load_crls(path)
     end
   end
 
+  # Ensures that the CA certificate is available for either generating or
+  # validating the host's cert.
+  # It will first check if the cert is present in memory (used for testing),
+  # then check on disk, and finally try to download it.
+  # @return true if the CA certificate was found, false otherwise
+  def ensure_ca_certificate
+    file_path = certificate_location(CA_NAME)
+    if check_for_certificate_in_memory(CA_NAME)
+      true
+    elsif Puppet::FileSystem.exist?(file_path)
+      begin
+        load_certificate_bundle(Puppet::FileSystem.read(file_path))
+        true
+      rescue ArgumentError
+        raise Puppet::Error, _("The CA certificate at %{file_path} is invalid Could not load.") % { file_path: file_path }
+      end
+    else
+      bundle = download_ca_certificate_bundle
+      if bundle
+        save_certificate_bundle(bundle)
+        true
+      else
+        false
+      end
+    end
+  end
+
+  # Creates an arry of SSL Certificate objects from a PEM-encoding string
+  # of one or more certs.
+  # @param [String] bundle_string PEM-encoded string of certs
+  # @return [[OpenSSL::X509::Certificate]] the certs loaded from the input string
+  def load_certificate_bundle(bundle_string)
+    delimiters = /-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----/m
+    certs = bundle_string.scan(delimiters)
+
+    if certs.empty?
+      raise ArgumentError, _("Not a valid certificate bundle")
```

Comment:
In current master, we raise and spit out a rather cryptic amalgamation of format and OpenSSL errors:
```
Error: Could not request certificate: Could not intern from text/plain: nested asn1 error
```
I think raising is better than returning `nil`, because if the CA cert is bad, we shouldn't try to continue (when I was just returning nil and warning, we tried to download the bad CA cert at least three times). I think it'd be best to use a different error type and catch that to emit a nicer message like the way I'm handling it here. In my manual testing, that definitely provided the nicest UX.

Reply:
Would it be reasonable to raise a `OpenSSL::X509::CertificateError`, do you think? The only reason we don't is because the `scan` returns an empty array if the string doesn't have the delimiters (which hides the case where the cert corruption is so bad that the response doesn't look like a cert). If you ask OpenSSL to parse _anything_ that's invalid, it will always raise that error, so that feels right to me, but it still falls under raising error types we don't own. Thoughts?

Reply:
I think I'm having a post lunch food coma, apologies if I'm not following what you said and am talking a bit around the point, but:

I still think its a good idea to error when we can't load the cert, but the usage of raising an exception in loading and then having callers catch and raise a different exception is what I was referring to. Like:
```rb
    elsif Puppet::FileSystem.exist?(file_path)
      begin
        load_certificate_bundle(Puppet::FileSystem.read(file_path))
        true
      rescue ArgumentError
        raise Puppet::Error, _("The CA certificate at %{file_path} is invalid Could not load.") % { file_path: file_path }
      end
```
in a caller would become:
```rb
    elsif Puppet::FileSystem.exist?(file_path)
      if load_certificate_bundle(Puppet::FileSystem.read(file_path))
        true
      else
        raise Puppet::Error, _("The CA certificate at %{file_path} is invalid Could not load.") % { file_path: file_path }
      end
```
I think that's an improvement, and an alright interface (whether or not trying to load invalid content is "exceptional" is something we could debate), but isn't a huge deal. I would say that if we keep it, we should change the exception to a different one though, as I do think catching `ArgumentError` will mask bugs in the future.

Reply:
That looks alright to me. I'll update it to that. We'll still need the one rescue to catch the OpenSSL error, but that can just return nil.